spacesuit/src/value.rs

```diff
@@ -126,10 +126,24 @@ impl Into<Scalar> for SignedInteger {
 }
 
 impl Add for SignedInteger {
-    type Output = SignedInteger;
+    type Output = Option<SignedInteger>;
 
-    fn add(self, rhs: SignedInteger) -> SignedInteger {
-        SignedInteger(self.0 + rhs.0)
+    fn add(self, rhs: SignedInteger) -> Option<SignedInteger> {
+        match self.0.checked_add(rhs.0) {
+            Some(x) => Some(SignedInteger(x)),
+            None => None,
+        }
+    }
+}
+
+impl Mul for SignedInteger {
+    type Output = Option<SignedInteger>;
+
+    fn mul(self, rhs: SignedInteger) -> Option<SignedInteger> {
+        match self.0.checked_mul(rhs.0) {
```

Comment:
Same as above, `checked_mul` is protecting against i128 overflows. However, `(2^64-1)^2` is larger than `2^127-1`, so we must use `checked_mul` here. But in addition to that we need to check that the resulting `x` is in good range ±(2^64-1) as we would do with `Add`.